### PR TITLE
pybullet envs: run IK once per option call and save a plan

### LINKED ISSUE #726: pybullet envs: run IK once per option call and memorize plan
this is the next step after #705 is closed, inching closer to #692

### PR DESCRIPTION
**Pull request by tomsilver**
closes #726

this PR leads to a small qualitative difference in the arm movements, see video: 
https://user-images.githubusercontent.com/2816502/163692996-a73b072f-7214-458d-a0e2-7351da21de7b.mp4

@ronuchit my code here is getting a little gross, would appreciate help on making it less gross (idea-wise or implementing it wise)

### CODE REVIEW

**Comment by tomsilver** [comment]
this PR is not ready for merge. there is performance degradation and failing seeds:
```
                              NUM_TRAIN_TASKS     NUM_SOLVED AVG_NUM_PREDS AVG_TEST_TIME  AVG_NODES_CREATED LEARNING_TIME  NUM_SEEDS
EXPERIMENT_ID                                                                                                                       
pybullet_blocks_nsrt_learning    50.00 (0.00)  43.87 (10.05)   5.00 (0.00)   0.39 (0.11)  3784.45 (1898.32)  22.79 (2.86)         15
pybullet_blocks_oracle            0.00 (0.00)   48.77 (0.58)   5.00 (0.00)   0.34 (0.08)  3747.56 (1464.68)   0.00 (0.00)         13
```

File: src/envs/pybullet_env.py [hunk 1]
```diff
@@ -329,37 +329,81 @@ def _create_move_end_effector_to_pose_option(
...
-                    params: Array) -> Action:
-            del memory  # unused
-            # First handle the main arm joints.
+        def _initiable(state: State, memory: Dict, objects: Sequence[Object],
+                       params: Array) -> bool:
+            # This is a really hacky hack to handle the situation where we are
+            # planning with a simplified model (that from the Blocks env) where
+            # states do not have joints. In this case, we want the option to
+            # be always initiable. Furthermore, we use pragma here because
+            # the oracle approach is not unit tested on this environment
```
Comment (ronuchit) [inline src/envs/pybullet_env.py:338]: you don't need to run the whole oracle approach to unit test this. you can just create a PyBulletBlocks object, and grab an option using `env.options`, then run `option.initiable()`

File: src/envs/pybullet_env.py [hunk 2]
```diff
@@ -329,37 +329,81 @@ def _create_move_end_effector_to_pose_option(
...
-            ee_action = np.add(current, ee_delta)
-            # We assume that the robot is already close enough to the target
-            # position that IK will succeed with one call, so validate is False.
-            # Furthermore, updating the state of the robot during simulation,
-            # which validate=True would do, is discouraged by PyBullet.
-            joint_state = self._pybullet_robot.run_inverse_kinematics(
-                (ee_action[0], ee_action[1], ee_action[2]), validate=False)
+            # Special case: we're already at the target. We should terminate
+            # successfully in one step.
+            if ee_norm == 0.0:
```
Comment (ronuchit) [inline src/envs/pybullet_env.py:359]: should this be a small threshold instead?

File: src/envs/pybullet_env.py [hunk 3]
```diff
@@ -329,37 +329,81 @@ def _create_move_end_effector_to_pose_option(
...
+                return True
+            # Extract the current joint state.
+            current_joints = cast(_PyBulletState, state).joint_state
+            # Make a plan in joint space.
             current, target, finger_status = \
                 get_current_and_target_pose_and_finger_status(
                     state, objects, params)
-            # Run IK to determine the target joint positions.
+            target_joints = self._pybullet_robot.run_inverse_kinematics(
+                target, validate=True)
```
Comment (ronuchit) [inline src/envs/pybullet_env.py:349]: why do we want validate=True here? remember the comment you deleted: "Furthermore, updating the state of the robot during simulation, which validate=True would do, is discouraged by PyBullet." so this seems like a bad thing to do?

File: src/envs/pybullet_env.py [hunk 4]
```diff
@@ -329,37 +329,81 @@ def _create_move_end_effector_to_pose_option(
...
+            finger_state = self._get_finger_state(state)
+            finger_status = memory["finger_status"]
+            # Determine if we have reached the next waypoint in the plan.
+            # The only way that the plan would be empty is if we were
+            # already close enough to the final waypoint, i.e. the target,
+            # in which case the option should have terminated and the
+            # policy should no longer be getting called.
+            waypoint = memory["waypoints"][0]
+            # The finger states should not be used in the distance computation
+            # because they will get overriden below.
```
Comment (ronuchit) [inline src/envs/pybullet_env.py:386]: below where?

File: src/envs/pybullet_env.py [hunk 5]
```diff
@@ -329,37 +329,81 @@ def _create_move_end_effector_to_pose_option(
...
+            # Extract the current finger state from the simulator state.
+            finger_state = self._get_finger_state(state)
+            finger_status = memory["finger_status"]
+            # Determine if we have reached the next waypoint in the plan.
+            # The only way that the plan would be empty is if we were
+            # already close enough to the final waypoint, i.e. the target,
+            # in which case the option should have terminated and the
+            # policy should no longer be getting called.
+            waypoint = memory["waypoints"][0]
+            # The finger states should not be used in the distance computation
```
Comment (ronuchit) [inline src/envs/pybullet_env.py:385]: how are they not being used? they're set to be part of the waypoints list:
```
waypoint[self._pybullet_robot.left_finger_joint_idx] = finger_state
waypoint[self._pybullet_robot.right_finger_joint_idx] = finger_state
```
and then used here: 
```
squared_dist = np.sum(np.square(np.subtract(current_joints, waypoint)))
```

**Review by ronuchit — COMMENTED** [review]
is it possible that changing to validate=True is what's screwing us up?

**Comment by tomsilver** [comment]
Resolution about this PR: we will close it, and instead, move the two main controllers into pybullet robots (so fetch can keep doing what it did before)

**Pull request closed without merging**